### PR TITLE
Server-side full text search with query parameters

### PR DESCRIPTION
**Pull request by liangyuanruo**
## Search feature

This PR migrates existing search functionality server-side and introduces query parameters for each of the search terms
- `s` for search text
- `c` for category
- `l` for location

A user may also share the URL directly with others which will automatically apply the filters e.g. https://freestuff.sg/?s=harry+potter&c=books&l=tanglin.

As the implementation uses the History API to maintain state, users may also use the back and forward buttons to navigate between searches.

### Debouncing search filter
To keep the number of network calls efficient, search is denounced to 300 milliseconds.

### Update listings by replacing the DOM subtree
As discussed, search listings are updated by parsing the templated HTML response and replacing the relevant DOM subtree with the new DOM elements.

## Database migration
Add the search index to the database via the following SQL commands:

```sql
ALTER TABLE listing
ADD COLUMN textsearchable_index_col tsvector
	GENERATED ALWAYS AS (
		to_tsvector('english', coalesce(listing_description, ''))
	) STORED;

CREATE INDEX textsearch_idx ON listing USING GIN (textsearchable_index_col);
```

## Other opportunities for improvement

It is possible to expand full-text search further over locations, categories, or even collection instructions, by building the search index over a concatenation of multiple table columns.

We may also wish to limit the number of maximum listings returned by the database via a LIMIT clause, or introduce pagination, if the number of listings grows too large.

### CODE REVIEW

**Comment by fynyky** [comment]
Client side stuff is good, but on the server side we need to do a db upgrade to support an index. Would also prefer not to introduce an ORM if I can avoid it.

**Comment by liangyuanruo** [comment]
> we need to do a db upgrade to support an index

OK, I will expand the scope of the PR to tackle this.

> Would also prefer not to introduce an ORM if I can avoid it.

Let me look into how to use Knex purely as a query builder, as this should be possible.

**Comment by fynyky** [comment]
Not quite what I meant. I'd prefer to drop Knex entirely and just use node postgres and construct the string. If this gets more complicated then sure, but for now this is the only use case so would prefer to avoid the additional dependency. Basically it involved building the string while tracking the number of params

**Comment by liangyuanruo** [comment]
@fynyky I see. Actually, my initial thoughts were to construct the SQL query manually as well. In the end, I decided to reach for a known solution to sidestep the problem quickly while keeping the code sane. I also thought that the abstraction provided by a query builder is relatively thin compared to a full-fledged ORM (for example, there is no mapping of DB records to classes or adding additional methods). 

Update: I have ported functionality out of Knex as requested.

**Comment by liangyuanruo** [comment]
OK I have implemented full-text search, let me experiment with constructing our own SQL queries

**Comment by liangyuanruo** [comment]
Merge conflicts have been resolved.

File: app/src/index.js [hunk 1]
```diff
@@ -141,36 +144,31 @@ app.use(express.static(publicPath))
...
-    // Logged in users see can their own posts even if less than 48 hours
-      result = await db.query(`
-      SELECT listing_id, listing_description, listing_location, listing_created_at, listing_image_key, listing_category
-      FROM listing JOIN account 
-      ON listing_owner_id = account_id
-      WHERE (listing_created_at < NOW() - INTERVAL '48 hours')
-      OR (listing_owner_id = $1)
-      ORDER BY listing_created_at DESC
-    `, [req.user.id])
+      ;({ sql, bindings } = getListingQuery('charity', { searchTerm, searchCategory, searchLocation }))
```
Comment (liangyuanruo) [inline app/src/index.js:159]: Change API parameter

File: app/src/js/index.js [hunk 2]
```diff
@@ -1,35 +1,89 @@
...
+    })
+    .catch(error => console.error(`fetchAndUpdateListings failed with error: ${error}`))
+}
+
+/**
+ * Function that returns an event handler that syncs an Event.target to a URL query parameter.
+ * @param {string} key of the query parameter to use
+ * @returns Event handler
+ */
+const onChange = (param) => (event) => {
```
Comment (liangyuanruo) [inline app/src/js/index.js:67]: rename to something with more explanatory value

**Pull request merged**